### ISSUE TITLE
Native PyTorch is faster than PyTorch Lightning (CPU) w/ Simple SGD Logistic Regression

### ISSUE THREAD

**Issue by bharathc346** [PyTorchLightning/pytorch-lightning#8857]
## 🐛 Bug

Pytorch Lightning is much slower than Native Pytorch

### To Reproduce

```py
import time
from tqdm import tqdm
from typing import Tuple, List, Dict, Optional

from joblib import Memory

import numpy as np
import pandas as pd
import scipy.sparse as sparse
import matplotlib.pyplot as plt
from sklearn.metrics import roc_auc_score
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import ParameterGrid


import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.utils.data as data
from torch import Tensor
from torch.optim.optimizer import Optimizer
from torchmetrics.functional import auroc

import pytorch_lightning as pl
from pytorch_lightning import seed_everything

from private_custom_library import ScipyCSRDataset

memory = Memory(".cached", verbose=0)

class LR(nn.Module):
    def __init__(self, feat_size: int):
        super().__init__()
        self.feat_size = feat_size
        self.linear = nn.Linear(feat_size, 1)

    def forward(self, x: Tensor):
        return self.linear(x)

class LightningLR(pl.LightningModule):
    def __init__(self, feat_size: int, learning_rate: float):
        super().__init__()
        self.feat_size = feat_size
        self.linear = nn.Linear(feat_size, 1)
        self.learning_rate = learning_rate

    def forward(self, x: Tensor):
        return self.linear(x)

    def training_step(
        self,
        batch: Tuple[Tensor, Tensor],
        batch_idx:  int
    ) -> Dict[str, Tensor]:
        x, y = batch
        loss = F.binary_cross_entropy_with_logits(
            self.linear(x), y
        )
        return {"loss": loss}

    def predict_step(
        self,
        batch: Tuple[Tensor, Tensor],
        batch_idx: int,
        dataloader_idx: Optional[int] = None
    ) -> Dict[str, float]:
        x, y = batch
        logits = self.linear(x)
        probs = torch.sigmoid(logits)
        return {
            "auc": auroc(probs, y.int()).item()
        }

    def configure_optimizers(self) -> Optimizer:
        return torch.optim.SGD(self.parameters(), lr=self.learning_rate)


class SparseTorchDataset(data.Dataset):
    def __init__(self, dataset: ScipyCSRDataset):
        self.X = self._process_X(dataset.X.tocoo())
        self.y = torch.from_numpy(dataset.y).float().unsqueeze(-1)

    def _process_X(self, X: sparse.coo_matrix):
        return torch.sparse.FloatTensor(
            torch.tensor([X.row.tolist(), X.col.tolist()], dtype=torch.long),
            torch.tensor(X.data, dtype=torch.float)
        )

    def __getitem__(self, idx: int) -> Tuple[Tensor, Tensor]:
        return self.X[idx], self.y[idx]

    def __len__(self) -> int:
        return self.y.shape[0]

class DenseTorchDataset(data.Dataset):
    def __init__(self, dataset: ScipyCSRDataset):
        self.X = dataset.X.toarray().astype(np.float32)
        self.y = torch.from_numpy(dataset.y).float().unsqueeze(-1)

    def __getitem__(self, idx: int) -> Tuple[Tensor, Tensor]:
        return self.X[idx], self.y[idx]

    def __len__(self) -> int:
        return self.y.shape[0]

@memory.cache
def get_datasets(feature_size: int) -> Tuple[ScipyCSRDataset, ScipyCSRDataset]:
    ...

def train_test_pure_torch(
    train: ScipyCSRDataset,
    test: ScipyCSRDataset,
    feature_size: int,
    sparse: bool = True,
    learning_rate: float = 1e-2,
    n_epochs: int = 1
) -> Tuple[float, float]:
    if sparse:
        train_loader = data.DataLoader(SparseTorchDataset(train), batch_size=1)
        test_loader = data.DataLoader(SparseTorchDataset(test),
            batch_size=len(test))
    else:
        train_loader = data.DataLoader(DenseTorchDataset(train), batch_size=1)
        test_loader = data.DataLoader(DenseTorchDataset(test),
            batch_size=len(test))

    model = LR(feature_size)
    optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

    start = time.time()
    for i in range(n_epochs):
        for batch in train_loader:
            x, y = batch

            optimizer.zero_grad()
            logits = model(x)
            loss = F.binary_cross_entropy_with_logits(logits, y)
            loss.backward()
            optimizer.step()
    end = time.time()

    all_probs, all_ys = [], []
    with torch.no_grad():
        for batch in test_loader:
            x, y = batch
            logits = model(x).squeeze(-1)
            all_probs.append(torch.sigmoid(logits))
            all_ys.append(y.squeeze(-1))
    probs = torch.cat(all_probs).numpy()
    y = torch.cat(all_ys).numpy().astype(int)
    auc = roc_auc_score(y, probs)
    return auc, end - start

def train_test_lightning(
    train: ScipyCSRDataset,
    test: ScipyCSRDataset,
    feature_size: int,
    sparse: bool = True,
    learning_rate: float = 1e-2,
    n_epochs: int = 1
) -> Tuple[float, float]:
    if sparse:
        train_loader = data.DataLoader(SparseTorchDataset(train), batch_size=1)
        test_loader = data.DataLoader(SparseTorchDataset(test),
            batch_size=len(test))
    else:
        train_loader = data.DataLoader(DenseTorchDataset(train), batch_size=1)
        test_loader = data.DataLoader(DenseTorchDataset(test),
            batch_size=len(test))

    model = LightningLR(feat_size=feature_size, learning_rate=learning_rate)
    trainer = pl.Trainer(
        max_epochs=n_epochs
    )
    start = time.time()
    trainer.fit(model, train_loader)
    end = time.time()
    auc = trainer.predict(dataloaders=test_loader).pop()["auc"]
    return auc, end - start

def train_test_sklearn(
    train: ScipyCSRDataset,
    test: ScipyCSRDataset,
    sparse: bool = True,
    learning_rate: float = 1e-2,
    n_epochs: int = 1,
    seed: Optional[int] = None
) -> Tuple[float, float]:
    if not sparse:
        train.X = train.X.toarray()
        test.X = test.X.toarray()

    clf = SGDClassifier(loss="log", alpha=0., max_iter=n_epochs,
        learning_rate="constant", eta0=learning_rate)
    start = time.time()
    clf.fit(train.X, train.y)
    end = time.time()
    auc = roc_auc_score(test.y, clf.predict_proba(test.X)[:, 1])
    return auc, end - start

def benchmark(
    n_trials: int = 5,
    feature_sizes: List[int] = [500, 1000],
    sparse_trials: List[bool] = [True, False],
) -> List[dict]:
    grid = list(ParameterGrid({
        "seed": list(range(n_trials)),
        "feature_size": feature_sizes,
        "sparse": sparse_trials,
    }))

    results = []
    for trial in tqdm(grid):
        pl.seed_everything(trial["seed"])
        train, test = get_datasets(trial["feature_size"])

        lightning_auc, lightning_time = train_test_lightning(
            train.copy(), test.copy(),
            feature_size=trial["feature_size"],
            sparse=trial["sparse"]
        )
        torch_auc, torch_time = train_test_pure_torch(
            train.copy(), test.copy(),
            feature_size=trial["feature_size"],
            sparse=trial["sparse"]
        )
        sklearn_auc, sklearn_time = train_test_sklearn(
            train.copy(), test.copy(),
            sparse=trial["sparse"]
        )

        results.append({
            "lightning_auc": lightning_auc,
            "lightning_time": lightning_time,
            "torch_auc": torch_auc,
            "torch_time": torch_time,
            "sklearn_auc": sklearn_auc,
            "sklearn_time": sklearn_time,
            **trial
        })
    return results

def plot_benchmarks(df: pd.DataFrame) -> None:
    ...

if __name__ == "__main__":
    results = benchmark()
    df = pd.DataFrame(results)
    df.to_csv("results.csv", index=False)
    df = pd.read_csv("results.csv")
    plot_benchmarks(df)

```

### Expected behavior

Believe pytorch_lightning should be faster or at least same speed as native pytorch with this simple model

### Environment
* CUDA:
	- GPU:
	- available:         False
	- version:           None
* Packages:
	- numpy:             1.21.1
	- pyTorch_debug:     False
	- pyTorch_version:   1.9.0
	- pytorch-lightning: 1.4.1
	- tqdm:              4.61.1
* System:
	- OS:                Darwin
	- architecture:
		- 64bit
	- processor:         i386
	- python:            3.8.0
	- version:           Darwin Kernel Version 20.6.0: Wed Jun 23 00:26:31 PDT 2021; root:xnu-7195.141.2~5/RELEASE_X86_64

### Additional context
Just trying to switch from using SGDClassifier in sklearn to a model in pytorch which allows sparse features (bag of words, tfidf) for similar performance in time and in auc. 

Here are the plots produced by the script above demonstrating that lightning is slower than native pytorch.

![dense](https://user-images.githubusercontent.com/53350528/129077459-cfeedee0-6edf-4e91-9bef-45425eb4e26d.png)
![sparse](https://user-images.githubusercontent.com/53350528/129077475-63d83258-f9cf-4dd8-97e0-c84209057334.png)

Note in the script `ScipyCSRDataset` is just a class that implements `__getitem__` like so
```
def __getitem__(self, idx): 
     # X is scipy.sparse.csr_matrix
     self.X[idx], self.y[idx]
```

**Comment by justusschock** [comment]
Hi @bharathc346 , thanks for opening the issue with such a good reproduction example :)

You are right, we are a bit slower, but this is a constant overhead, which comes from internal preparations for additional features as logging and some more checks in the background. This however, should be a constant offset (we benchmark it ourselves in the CI for CNNs and RNNs and there we meet the speed except for a constant offset which is really minor for NN workloads, but unfortunately, you are experiencing it quite large since your epoch time is so low.

**Comment by bharathc346** [comment]
Hi @justusschock. This makes sense, are these internal preparations being done for epoch? or is this something being done at the beginning of training. I guess what I'm asking is should the second and onward epochs of training in Pytorch vs Pytorch Lightning be around similar speeds?

Will benchmark this later today if I have some time.

**Comment by NumberChiffre** [comment]
@bharathc346 Same issues here [#6196](https://github.com/PyTorchLightning/pytorch-lightning/issues/6196) and [#7417](https://github.com/PyTorchLightning/pytorch-lightning/issues/7417)

Let us know if you are doing more benchmarks on larger dataset/neural nets!

**Comment by justusschock** [comment]
Hi @bharathc346 ,

Some of them are, some of them unfortunately not. E.g. everything related to data/logging has to be done every epoch, but setting up profiling and device selection/many other sanity checks are part of the first epoch.

You can find our parity tests, where we compare against pure pytorch in speed and accuracy [here](https://github.com/PyTorchLightning/pytorch-lightning/tree/master/benchmarks). Hope that helps. Would be interested in your insights :)

**Comment by adamgayoso** [comment]
Is there any benchmarking done to see how the speed has changed over pytorch lightning versions? Anecdotally we just upgraded to >1.4 and it appears to be slower than 1.3.x.

**Comment by justusschock** [comment]
@adamgayoso do you have a reproducible script for us so that we can look into this?

There is no benchmarking done across different PL versions, but we run benchmarks against pure PyTorch and ensure that for NNs we hit the same runtime

**Comment by adamgayoso** [comment]
Please see #9304
